### ISSUE TITLE
I2P: no outgoing connections to peers after resolve is complete

### ISSUE THREAD

**Issue by Vort** [arvidn/libtorrent#7262]
There are two similar functions in `torrent.cpp`: `on_peer_name_lookup` and `on_i2p_resolve`.
One of them, `on_peer_name_lookup` contains `update_want_peers();` call at the end:
https://github.com/arvidn/libtorrent/blob/b17f6150ecb0cf5b335305124a4400fc81a03831/src/torrent.cpp#L3795
But other one, `on_i2p_resolve`, do not have it:
https://github.com/arvidn/libtorrent/blob/b17f6150ecb0cf5b335305124a4400fc81a03831/src/torrent.cpp#L3744
It looks like a bug.
When I added `update_want_peers();` call at the end of `on_i2p_resolve` function, my test program started working as expected: lots of outgoing connections start to spawn and download start going.

Related: https://github.com/arvidn/libtorrent/issues/2530

_libtorrent version (or branch)_: 690881fec9cc069976cd3d909254a87c22f43ce0 + fixes from #7247, #7248, #7259 and #7260
_platform/architecture_: Windows x64
_compiler and compiler version_: clang 14.0.0

**Comment by jiigen** [comment]
Going to test these changes  and will share my result. Since I use a different OS and compiler could be useful I guess.

**Comment by jiigen** [comment]
No luck this time...Added both your changes to the code and recompiled but I only seen the  naming lookup. After that it get stuck as before and no download nor connection to peers.
To be clear I've added: _update_want_peers();_  after _peers_erased(st.erased);_

**Comment by Vort** [comment]
> To be clear I've added: update_want_peers(); after peers_erased(st.erased);

Right, at line # 3744.

> Added both your changes to the code and recompiled but I only seen the naming lookup.

To diagnose problem you can replace `lt::alert_category::status | lt::alert_category::error` with `lt::alert::all_categories` in test program.

If output have `I2P+ Universal Installer 2.0.0+ (http://tracker2.postman.i2p/announce.php)[0.0.0.0:6881] v1 received peers: 37`, then it confirms that previous fix (peer list / naming lookup) is working.
If after that several `I2P+ Universal Installer 2.0.0+: i2p_resolve error: invalid key` appear, then it is fine - for some reason not every name resolves correctly.

Without `update_want_peers();` nothing happens next.
But with it, you should see `I2P+ Universal Installer 2.0.0+: *** UPDATE LIST [ torrent_want_peers_download : 1 ]` and `I2P+ Universal Installer 2.0.0+ peer [ 0.0.0.0:0 client: Unknown ] [0.0.0.0:0] >>> OUTGOING_CONNECTION [ ep: 0.0.0.0:0 type: I2P seed: 0 p: 000000000222D430 local: 0.0.0.0:0 ]`.

**Comment by jiigen** [comment]
Obviously the output got very verbose but yes I see many attempt to connect to peers even if so far all of them get refused but I'm still investigating.
Too bad I wasn't able to build python bindings (for several reasons) but I think that if you manage to build python bindings you can do tests even with a "real" client: Deluge.
Deluge has i2p options in the interface and has and extesion which let you set other paramaters of libtorrent which is used as backend.

**Comment by Vort** [comment]
Just to be sure: you see `received peers: 37` (or similar number, but not 200 or 300), `torrent_want_peers_download : 1` and `OUTGOING_CONNECTION [ ep: 0.0.0.0:0 type: I2P`?

**Comment by jiigen** [comment]
Yes i get:
received peers: 186
And several:
I2P+ Universal Installer 2.0.0+ peer [ ip:port] port client: Unknown ] [ip:port] >>> OUTGOING_CONNECTION [ ep: ip:port type: I2P seed: 0 p: 0x7f1b04015760 local: 0.0.0.0:0 ]

**Comment by jiigen** [comment]
Yes i get:
received peers: 186
And several:
I2P+ Universal Installer 2.0.0+ peer [ ip:port] port client: Unknown ] [ip:port] >>> OUTGOING_CONNECTION [ ep: ip:port type: I2P seed: 0 p: 0x7f1b04015760 local: 0.0.0.0:0 ]

**Comment by Vort** [comment]
> i get:
> received peers: 186

> ep: ip:port

This is wrong.
It means that fix from #7260 is not working.

**Comment by jiigen** [comment]
What's wrong with that?

**Comment by jiigen** [comment]
I2P+ Universal Installer 2.0.0+: *** UPDATE LIST [ torrent_want_peers_download : 1 ]
This is another line i get in the log

**Comment by Vort** [comment]
> What's wrong with that?

in Compact Tracker Response (https://geti2p.net/en/docs/applications/bittorrent) `peers` key consists of 32-byte hashes.
If it becomes parsed as IPv4 (4 byte size) instead, then incorrect peer count is displayed (8 times more) and you see (fake) IP addresses in `ep:` instead of `0.0.0.0`.
This problem should be fixed with #7260, but for some reason this fix does not work for you.

**Comment by jiigen** [comment]
Will check the code later to see if i did sone typos but I'm pretty sure it should be correct

**Comment by jiigen** [comment]
Why you dont make a "i2p-fixes" branch on your libtorrent fork ?

**Comment by Vort** [comment]
> Why you dont make a "i2p-fixes" branch on your libtorrent fork ?

1. Developers usually better know how exactly bugs should be fixed. I don't want my potential mistakes to go into release with a simple merge.
2. I need hacks to make this build work for me (first one is #6988, second one is I need to split `http_connection.cpp` in two parts, otherwise Windows debug build fails).

If that is not a problem, I can share it of course: https://github.com/Vort/libtorrent/commit/622bcce3cc1c2b7c45fdd62603c5072e9c4f8523

**Comment by jiigen** [comment]
Ok all done, maybe I messed up something when linking against the library. I redid everything from scratch and not it works!
Peers get connected correctly and download started very quick. 
I even checked the i2pd web console  while downloading and it shows:

```
session [127.0.0.1:53464]
acceptor [127.0.0.1:47412]
stream [127.0.0.1:39082]
stream [127.0.0.1:39092]
stream [127.0.0.1:39100]
stream [127.0.0.1:39114]
stream [127.0.0.1:39116]
stream [127.0.0.1:39120]
stream [127.0.0.1:32796]
stream [127.0.0.1:32808]
stream [127.0.0.1:32814]
stream [127.0.0.1:32822]
stream [127.0.0.1:38346]
stream [127.0.0.1:38354]
........................................
```
All seems to work pretty good so far. Thank you very much!

**Comment by jiigen** [comment]
Managed to build the python bindings and it works even with Deluge!
Here's some screenshots:
https://postimg.cc/gallery/NB6DFG5

**Comment by jiigen** [comment]
UPDATE:
I got peak speed of 180Kb/s which are quite aligned with I2psnark speeds with 3hop tunnels.
So I guess that when will be possible to set tunnel lenght and using less "safe" 2hop tunnels the speed should increase.

**Comment by Vort** [comment]
I made test download also with [client_test](https://github.com/arvidn/libtorrent/blob/RC_2_0/examples/client_test.cpp).
I launched it with `run.bat` file in Windows with such contents:
`client_test --enable_dht=false --proxy_type=i2p_proxy --i2p_hostname=127.0.0.1 --i2p_port=7656 magnet:?xt=urn:btih:cb18974db461c640c0d938710203bb263a746694^&dn=I2P%%2B+Universal+Installer+2.0.0%%2B^&tr=http://tracker2.postman.i2p/announce.php`
Not the fastest tunnel was selected this time, but download works:
![i2p_bt_test](https://user-images.githubusercontent.com/1242858/212659666-7cf9652f-b90c-45c8-bf54-09879c9fb132.png)

**Comment by jiigen** [comment]
I forgot...with the setting above I got peak speeds on some torrents of more than 500Kb/s with something like 12/13 peers.

**Comment by Vort** [comment]
> if this will bring other bugs

I think no.

> I noticed a slight reduction of the time required for the SAM acceptor creation

Because tunnel creation can fail.
And longer tunnel have higher chance of failure.
But there is downside too: `.length=1` make connections less secure.
I think it should be set to default value of `3`, but be allowed to changed with options by user.

---

I tested your change and I confirm that it works.
Also with shorter tunnels there are higher chances to have fast download speeds.
With this change I saw 100 kB/s compared to 30 kB/s in previous test.

**Comment by jiigen** [comment]
> > if this will bring other bugs
> 
> I think no.
> 
> > I noticed a slight reduction of the time required for the SAM acceptor creation
> 
> Because tunnel creation can fail. And longer tunnel have higher chance of failure. But there is downside too: `.length=1` make connections less secure. I think it should be set to default value of `3`, but be allowed to changed with options by user.
> 
> I tested your change and I confirm that it works. Also with shorter tunnels there are higher chances to have fast download speeds. With this change I saw 100 kB/s compared to 30 kB/s in previous test.

Oh yes I know that 1hop tunnels are not safe I put those on 1 to test the speed but without putting them to 0 which makes no sense in terms of privacy. I think both parameters should be available to applications through options when setting the i2p proxy parameters...

**Comment by jiigen** [comment]
I'm not a programmer I know C language but at an amateur/hobbist level and reading modern C++ is a pain for me but maybe you could be able to add some "lt::settings_pack" to set those parameters. I'v not those skills :P

**Comment by Vort** [comment]
> maybe you could be able to add some "lt::settings_pack" to set those parameters

I think for now it is enough to set `.length` to `2` or `3` and that's it.
Here is how it is done in i2psnark: https://github.com/i2p/i2p.i2p/blob/eb03e19f03d4a2dbb5bca776bc02583da53c88f1/apps/i2psnark/java/src/org/klomp/snark/SnarkManager.java#L840-L842

Working I2P support is number one priority.
Perfect I2P support is second priority and can be done after "working" state is achieved (and in other issues/PRs).

**Comment by jiigen** [comment]
> Working I2P support is number one priority. Perfect I2P support is second priority and can be done after "working" state is achieved (and in other issues/PRs).

So in your opinion what is still missing to say that "is working" after your fixes??

**Comment by Vort** [comment]
> So in your opinion what is still missing to say that "is working" after your fixes??

Changes should be checked by developers of both library and clients.
Also I encountered crash / failed assert [here](https://github.com/arvidn/libtorrent/blob/4d65bc33426a525aef92ac3ccdeb8bb3de1084b1/src/request_blocks.cpp#L146). Not sure if it is important and where it should be fixed.

Basically I mean that it makes sense to continue polishing of I2P suport once these fixes are implemented in main branch of libtorrent.

**Comment by arvidn** [comment]
@Vort would you mind making a PR?

**Comment by Vort** [comment]
> would you mind making a PR?

Ok, I made it: #7272

**Comment by zzzi2p** [comment]
don't set length to 1 or 2. 3 is both the min and the max required for anonymity, and it is the default in both routers.

Ideally, you'd provide a standard way for downstreams to set these configuration options

**Comment by Vort** [comment]
I selected 3 in PR. Because of i2psnark.

**Comment by jiigen** [comment]
> > would you mind making a PR?
> 
> Ok, I made it: #7272

I see you included my changes too, So i guess they were correct. I'm very glad.

**Comment by absolutep** [comment]
I have been requesting i2p feature across all torrent clients for more than an year now. Great to see it is finally happening.
Thanks to everyone who is trying to implement this feature.

**Comment by zzzi2p** [comment]
The PR looks good to me, nice job Vort and Jigen.

re: making the SAM settings configurable, that's probably best implemented afterward by the libtorrent team.
What I recommend providing for config is similar to what we do in i2psnark and will be familiar to i2p users and devs:

Four separate configs for tunnel inbound/outbound length/quantity (all integers, all default 3)
And a fifth config "other I2P options" that takes a freeform a=b c=d e=f string that you stuff into the SESSION CREATE line if set.

This will provide maximum flexibility to downstream devs and their users.

**Comment by Vort** [comment]
Fixed with #7272.
